python/shotgun_fields/image_widget.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Widget that represents the value of an image field in Shotgun
+"""
+import sgtk
+from sgtk.platform.qt import QtCore, QtGui
+from .widget_metaclass import ShotgunFieldMeta
+
+shotgun_data = sgtk.platform.import_framework("tk-framework-shotgunutils", "shotgun_data")
+
+
+class ImageWidget(QtGui.QLabel):
+    """
+    Inherited from a :class:`~PySide.QtGui.QLabel`, this class is able to
+    display an image field value as returned by the Shotgun API.
+    """
+    __metaclass__ = ShotgunFieldMeta
+    _FIELD_TYPE = "image"
+
+    def setup_widget(self):
+        """
+        Initialize the widget state.  Start up a Shotgun data retriever to download
+        images in the background.
+        """
+        self._pixmap = None
+
+        # start up a data retriever to fetch the thumbnail in the background
+        self._data_retriever = shotgun_data.ShotgunDataRetriever(bg_task_manager=self._bg_task_manager)
+        self._data_retriever.start()
+        self._data_retriever.work_completed.connect(self._on_worker_signal)
+        self._data_retriever.work_failure.connect(self._on_worker_failure)
+
+    def _display_default(self):
+        """ Default widget state is empty. """
+        self.clear()
+
+    def _display_value(self, value):
+        """
+        Set the value displayed by the widget.
+
+        :param value: The value displayed by the widget
+        :type value: A String that is a valid URL for the thumbnail image
+        """
+        # queue up the download in the background
+        entity_id = self._entity and self._entity.get("id") or None
+        entity_type = self._entity and self._entity.get("type") or None
+        self._task_uid = self._data_retriever.request_thumbnail(
```

Comment:
if you pass a `request_thumbnail=True` parameter here, the data dictionary you get back when the worker signals will contain a ready-loaded `QImage` instance of the thumbnail. (docs here: http://developer.shotgunsoftware.com/tk-framework-shotgunutils/shotgun_data.html#shotgun_data.ShotgunDataRetriever.request_thumbnail ). This might be nice because the image is then loaded from disk and processed in the worker thread instead of in the main one!